File: app/Models/DirectoryWatcher.swift
import Foundation

/// Per-directory kqueue watching, ported from file_explorer_swift's `LocalFileSource.watch`.
///
/// Not FSEvents: its coalescing latency is 1-3 seconds, which reads as lag when you save in
/// another editor and tab back. md-boss only ever watches the small bounded set of expanded
/// directories, which is exactly the granularity kqueue is good at.
@MainActor
final class DirectoryWatcher {
    enum Event {
        /// Something in the directory changed - re-list it.
        case changed
        /// The directory itself was deleted or renamed out from under us.
        case vanished
    }

    /// O_EVTONLY descriptors are real file descriptors. A runaway expansion would exhaust
    /// the process limit, so watching stops here and the UI says so.
    static let maxWatchers = 128

    private var entries: [String: Entry] = [:]
    private let onEvent: (URL, Event) -> Void

    private(set) var isSaturated = false

    init(onEvent: @escaping (URL, Event) -> Void) {
        self.onEvent = onEvent
    }

    deinit {
        // Entry.deinit cancels each source, which closes the descriptor.
    }

    var count: Int { entries.count }

    // MARK: Membership

    /// Watches exactly `urls` - adds what is new, drops what is gone. Called whenever the
    /// set of expanded directories changes.
    func sync(to urls: Set<URL>) {
        let wanted = Set(urls.map(\.path))

        for path in entries.keys where !wanted.contains(path) {
            entries[path] = nil
        }

        for url in urls where entries[url.path] == nil {
            guard entries.count < Self.maxWatchers else {
                isSaturated = true
                return
            }
            entries[url.path] = makeEntry(for: url)
        }

        isSaturated = false
    }

    func unwatchAll() {
        entries.removeAll()
        isSaturated = false
    }

    /// Atomic writes replace the inode via rename(2), so the descriptor we hold is left
    /// watching a deleted file. Every save path must call this or external-change detection
    /// silently stops working after the first save.
    func rearm(_ url: URL) {
        guard entries[url.path] != nil else { return }
        entries[url.path] = makeEntry(for: url)
    }

    // MARK: Internals

    private func makeEntry(for url: URL) -> Entry? {
        let descriptor = open(url.path, O_EVTONLY)
        guard descriptor >= 0 else { return nil }

        let source = DispatchSource.makeFileSystemObjectSource(
            fileDescriptor: descriptor,
            eventMask: [.delete, .rename, .write, .extend],
            queue: .main
        )

        source.setEventHandler { [weak self] in
            let flags = source.data
            let event: Event = flags.contains(.delete) || flags.contains(.rename) ? .vanished : .changed
            MainActor.assumeIsolated { self?.dispatch(url, event) }
        }
        source.setCancelHandler { close(descriptor) }
        source.resume()

        return Entry(source: source)
    }

    /// A `git checkout` or a build tool fires dozens of events in a burst; re-listing on
    /// each one would make the tree flicker and lose the cursor.
    private func dispatch(_ url: URL, _ event: Event) {
        guard let entry = entries[url.path] else { return }

        if event == .vanished {
            entries[url.path] = nil
            onEvent(url, .vanished)
            return
        }

        entry.debounce?.cancel()
        entry.debounce = Task { [weak self] in
            try? await Task.sleep(nanoseconds: 200_000_000)
            guard !Task.isCancelled else { return }
            self?.onEvent(url, .changed)
        }
    }

    private final class Entry {
        let source: DispatchSourceFileSystemObject
        var debounce: Task<Void, Never>?

        init(source: DispatchSourceFileSystemObject) {
            self.source = source
        }

        deinit {
            debounce?.cancel()
            source.cancel()
        }
    }
}

File: src/models/directoryWatcher.ts
// Per-directory watching of a small bounded set - the active root and the expanded folders,
// which is exactly the granularity a non-recursive watch is good at on every platform. A
// runaway expansion would exhaust inotify or descriptor limits, so watching stops at a cap
// and the UI can say so.

import { native, type Unwatch } from '../native/bridge'

export class DirectoryWatcher {
  static readonly maxWatchers = 128

  private readonly entries = new Map<string, Promise<Unwatch>>()
  private readonly onChange: (dir: string, changed: string[]) => void
  isSaturated = false

  constructor(onChange: (dir: string, changed: string[]) => void) {
    this.onChange = onChange
  }

  get count(): number {
    return this.entries.size
  }

  /** Watches exactly `dirs` - adds what is new, drops what is gone. */
  sync(dirs: ReadonlySet<string>): void {
    for (const [dir, pending] of this.entries) {
      if (dirs.has(dir)) continue
      this.entries.delete(dir)
      void pending.then((unwatch) => unwatch()).catch(() => {})
    }
    for (const dir of dirs) {
      if (this.entries.has(dir)) continue
      if (this.entries.size >= DirectoryWatcher.maxWatchers) {
        this.isSaturated = true
        return
      }
      const pending = native().watch(dir, (changed) => {
        // A late event from a watch that was dropped meanwhile is noise.
        if (this.entries.has(dir)) this.onChange(dir, changed)
      })
      // A folder that cannot be watched - gone, denied - is simply not watched; the poll
      // still covers it. Said out loud, because a silent watcher is indistinguishable from
      // a working one until a file fails to appear.
      this.entries.set(dir, pending.catch((err: unknown) => {
        console.warn(`watch failed for ${dir}:`, err)
        return () => {}
      }))
    }
    this.isSaturated = false
  }

  unwatchAll(): void {
    this.sync(new Set())
  }
}
